pkg/service/model.go

```diff
@@ -41,50 +110,206 @@ func (s *Service) ConvertToMDModel(ctx context.Context, fileUID uuid.UUID, calle
 	// Get the appropriate prefix for the file type
 	prefix := getFileTypePrefix(fileType)
 
+	// Prepare files to process (as a slice of base64 strings)
+	var filesToProcess []string
+	var err error
+
+	// For PDFs, split into batches; for other file types, process as a single item
+	if fileType == artifactpb.FileType_FILE_TYPE_PDF {
+		pdfData, err := base64.StdEncoding.DecodeString(fileBase64)
+		if err != nil {
+			return "", fmt.Errorf("decoding base64 PDF: %w", err)
+		}
+		filesToProcess, err = splitPDFIntoBatches(pdfData, 5)
+		if err != nil {
+			return "", fmt.Errorf("splitting PDF into batches: %w", err)
+		}
+	} else {
+		filesToProcess = []string{fileBase64}
+	}
+
+	// Process all files (or batches for PDFs)
+	results := make([]string, len(filesToProcess))
+	errCh := make(chan error, len(filesToProcess))
+	var wg sync.WaitGroup
+
+	// Add a done channel to handle context cancellation
+	done := make(chan struct{})
+	defer close(done)
+
 	namespaceID := "admin"
-	req := &modelpb.TriggerNamespaceModelRequest{
-		NamespaceId: namespaceID,
-		ModelId:     ConvertDocToMDModelID,
-		Version:     ConvertDocToMDModelVersion,
-		TaskInputs: []*structpb.Struct{
-			{
-				Fields: map[string]*structpb.Value{
-					"data": {Kind: &structpb.Value_StructValue{
-						StructValue: &structpb.Struct{
-							Fields: map[string]*structpb.Value{
-								"doc_content": {Kind: &structpb.Value_StringValue{StringValue: prefix + fileBase64}},
-							},
+
+	// Start a goroutine to watch for context cancellation
+	go func() {
+		select {
+		case <-ctx.Done():
+			// If context is cancelled, put the error in the error channel
+			errCh <- ctx.Err()
+		case <-done:
+			// Operation completed normally
+			return
+		}
+	}()
+
+	for i, b64 := range filesToProcess {
+		wg.Add(1)
+		go func(idx int, b64 string) {
+			defer wg.Done()
+
+			select {
+			case <-ctx.Done():
+				// If parent context is already cancelled, don't proceed
+				errCh <- ctx.Err()
+				return
+			default:
+				// Continue with processing
+			}
+
+			// Increase the timeout for larger files (optional)
+			triggerCtx, cancel := context.WithTimeout(ctx, 60*time.Second)
+			defer cancel()
+
+			req := &modelpb.TriggerNamespaceModelRequest{
+				NamespaceId: namespaceID,
+				ModelId:     ConvertDocToMDModelID,
+				Version:     ConvertDocToMDModelVersion,
+				TaskInputs: []*structpb.Struct{
+					{
+						Fields: map[string]*structpb.Value{
+							"data": {Kind: &structpb.Value_StructValue{
+								StructValue: &structpb.Struct{
+									Fields: map[string]*structpb.Value{
+										"doc_content": {Kind: &structpb.Value_StringValue{StringValue: prefix + b64}},
+									},
+								},
+							}},
 						},
-					}},
+					},
 				},
-			},
-		},
+			}
+
+			// Retry mechanism parameters
+			maxRetries := 3
```

Comment:
I'm not sure if this is the best place to introduce a retrial mechanism. We can define a retrial policy in the model client [here](https://github.com/instill-ai/artifact-backend/blob/main/pkg/client/grpc/model.go#L34), which is perhaps cleaner.

If you want to do this, you can implement the feature without retrials and I can open a separate PR to include that